internal/endpoints/network.go

```diff
@@ -121,14 +126,24 @@ func (n *Network) Serve() {
 	}()
 }
 
-// Close the listener.
+// Close the listener and server.
+// Will attempt to close the server gracefully, if configured with a drain connections timeout.
+// Note that graceful shutdown will timeout if the connections do not finish (e.g.: a request caused the server
+// to Close the endpoints on the same goroutine).
 func (n *Network) Close() error {
 	if n.listener == nil {
 		return nil
 	}
 
 	logger.Info("Stopping REST API handler - closing https socket", logger.Ctx{"address": n.listener.Addr()})
-	n.cancel()
+	defer n.cancel()
```

Comment:
Yes. `n.cancel` refers to `n.ctx`'s cancel, but we're also passing `n.ctx` to `shutdownServer` below, which we'll use to gracefully shutdown the server. If we're canceling the context now, `server.Shutdown` will error out because the context was canceled:

```
Dec 11 22:53:41 ubuntu k8s.k8sd[162622]: time="2024-12-11T22:53:41Z" level=error msg="Failed to gracefully shutdown server" err="context canceled"
```

Reply:
Mh but the `defer` still gets executed before `shutdownServer` gets called below which means the context is already cancelled.

What about doing something like this:

```go
err = shutdownServer(n.ctx, n.server, n.drainConnectionsTimeout)
if err != nil {
  //
}

n.cancel()
```

This way you ensure the context is still intact when triggering the shutdown.